compiler/resolution.common/src/org/jetbrains/kotlin/resolve/calls/inference/model/MutableConstraintStorage.kt

```diff
@@ -60,7 +60,12 @@ class MutableVariableWithConstraints private constructor(
                 && previousConstraint.isNullabilityConstraint == constraint.isNullabilityConstraint
             ) {
                 if (newConstraintIsUseless(previousConstraint, constraint)) {
-                    return previousConstraint to false
+                    // Preserve constraints with different custom type attributes.
+                    // This allows us to union type attributes in NewCommonSuperTypeCalculator.kt
+                    with(context) {
+                        if (previousConstraint.type.getCustomAttributes() == constraint.type.getCustomAttributes())
```

Comment:
Maybe it's better to replace `getCustomAttributes` with `hasCustomAttributes` and only then filter attributes?